### PR TITLE
ObjectId use uint64 to reduce calls

### PR DESCRIPTION
**Pull request by gerhardol**
## Proposed changes

Store ObjectId in 2*uint64+1*uint32 instead of uint32 
This reduces the number of calls handling ObjectId and improves the performance slightly.

In my tests I parsed 100K references in 87ms instead of 110 ms, so loading a repo with the default 100K setting is about 70 ms faster.

See also #11222 that removes `bool Equals(string? other)` that adds duplicated calls for uint32/uint64
Further optimizations will follow, for instance for `string ToShortString(int length = 8)` reduce the code some.
(In general, the code size is reduced by this PR, only for these two methods that adds code)

## Test methodology 

Tests exists already

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by RussKie — APPROVED** [review]

File: Plugins/GitUIPluginInterfaces/ObjectId.cs [hunk 1]
```diff
@@ -320,27 +301,47 @@ public unsafe string ToShortString(int length = 8)
...
+                *p++ = ParseHexDigit(i >> 28);
+                *p++ = ParseHexDigit((i >> 24) & 0xF);
+                *p++ = ParseHexDigit((i >> 20) & 0xF);
+                *p++ = ParseHexDigit((i >> 16) & 0xF);
+                *p++ = ParseHexDigit((i >> 12) & 0xF);
+                *p++ = ParseHexDigit((i >> 8) & 0xF);
+                *p++ = ParseHexDigit((i >> 4) & 0xF);
+                *p++ = ParseHexDigit(i & 0xF);
+
+                char ParseHexDigit(uint j) => j < 10 ? (char)('0' + j) : (char)(j + 0x57);
```
Comment (mstv) [inline Plugins/GitUIPluginInterfaces/ObjectId.cs:343]:
```suggestion
                return;

                static char GetHexChar(uint nibble) => nibble < 10 ? (char)('0' + nibble) : (char)(('a' - 10) + nibble);
```

File: Plugins/GitUIPluginInterfaces/ObjectId.cs [hunk 2]
```diff
@@ -320,27 +301,47 @@ public unsafe string ToShortString(int length = 8)
...
-            Write(_i5);
+            Write64(_i1);
+            Write64(_i2);
+            Write32(_i3);
 
             return new string(buffer, 0, length);
 
-            void Write(uint i)
+            void Write64(ulong i)
             {
```
Comment (mstv) [inline Plugins/GitUIPluginInterfaces/ObjectId.cs:311]: ❔
```suggestion
            void Write64(ulong i)
            {
                Write32(i >> 32);
                Write32(i & 0xFFFF_FFFF);
            }
```

File: Plugins/GitUIPluginInterfaces/ObjectId.cs [hunk 3]
```diff
@@ -373,13 +372,38 @@ public bool Equals(string? other)
...
+
+                bool TestDigit(ulong j)
+                {
+                    char c = j < 10 ? (char)('0' + j) : (char)(j + 0x57);
+                    return other[i++] == c;
+                }
+            }
 
-            bool TestInt(uint k)
+            bool TestInt32(uint k)
```
Comment (mstv) [inline Plugins/GitUIPluginInterfaces/ObjectId.cs:406]: ditto

File: Plugins/GitUIPluginInterfaces/ObjectId.cs [hunk 4]
```diff
@@ -21,17 +21,17 @@ public sealed class ObjectId : IEquatable<ObjectId>, IComparable<ObjectId>
         /// <summary>
         /// Gets the artificial ObjectId used to represent working directory tree (unstaged) changes.
         /// </summary>
-        public static ObjectId WorkTreeId { get; } = new(0x11111111, 0x11111111, 0x11111111, 0x11111111, 0x11111111);
+        public static ObjectId WorkTreeId { get; } = new(0x1111111111111111, 0x1111111111111111, 0x11111111);
```
Comment (mstv) [inline Plugins/GitUIPluginInterfaces/ObjectId.cs:24]:
```suggestion
        public static ObjectId WorkTreeId { get; } = new(0x11111111_11111111, 0x11111111_11111111, 0x11111111);
```

File: Plugins/GitUIPluginInterfaces/ObjectId.cs (hunk 1 above)
Comment (mstv) [reply Plugins/GitUIPluginInterfaces/ObjectId.cs:343]: Will be replaced by #11224 anyway - but only if #11224 takes comparable runtime. Does it?

Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/ObjectId.cs:343]: 100K run of ToString() requires 13ms so less than a micro second (all are the same in the test though but do not expect much more). An additional allocation too. So nothing, even less than using LYNQ compared to direct. (It is the operations that is done 300K reading loading a repo that matters.)

(I had first planned to optimize that method but there is no benefit, just more code. It is too complicated with this PR, but if the others are merged first, we will not see it.)

File: Plugins/GitUIPluginInterfaces/ObjectId.cs (hunk 2 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/ObjectId.cs:311]: As #11222 was accepted, I will remove this code

File: Plugins/GitUIPluginInterfaces/ObjectId.cs (hunk 3 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/ObjectId.cs:406]: As #11222 was accepted, I will remove this code

File: Plugins/GitUIPluginInterfaces/ObjectId.cs (hunk 4 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/ObjectId.cs:24]: OK in groups with 8 or 4 digits?
(did not know about this feature)

Comment (mstv) [reply Plugins/GitUIPluginInterfaces/ObjectId.cs:24]: > in groups with 8 or 4 digits?

As you prefer.

Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/ObjectId.cs:24]: group by 4

**Review by mstv — APPROVED** [review]
LGTM, have not run

**Pull request merged**